Revision of xcube serve documentation (#710)
dcs4cop/xcube: docs/source/cli/xcube_serve.rst

                               paths in CONFIG and relative CUBES paths. Defaults
                               to value of environment variable
                               XCUBE_SERVE_BASE_DIR, if given, otherwise defaults
                               to the parent directory of CONFIG.
      --tilecache SIZE         In-memory tile cache size in bytes. Unit suffixes
                               'K', 'M', 'G' may be used. Defaults to '512M'. The
                               special value 'OFF' disables tile caching.
      --tilemode MODE          Tile computation mode. This is an internal option
                               used to switch between different tile computation
                               implementations. Defaults to 0.
      -s, --show               Run viewer app. Requires setting the environment
                               variable XCUBE_VIEWER_PATH to a valid xcube-viewer
                               deployment or build directory. Refer to
                               https://github.com/dcs4cop/xcube-viewer for more
                               information.
      -q, --quiet              Disable output of log messages to the console

>                                entirely. Note, this will also suppress error and

Author: This is the output of xcube serve --help, I will not change it here.

                               warning messages.
      -v, --verbose            Enable output of log messages to the console. Has
                               no effect if --quiet/-q is used. May be given
                               multiple times to control the level of log

>                                messages, i.e., -v refers to level INFO, -vv to

Author: This is the output of xcube serve --help, I will not change it here.

      - Identifier: my_dataset
        Title: "My Dataset"
        FileSystem: file
        Path: my_dataset.zarr
        TimeSeriesDataset: my_dataset_opt_for_ts

      - Identifier: my_dataset_opt_for_ts
        Title: "My Dataset optimized for Time-Series"
        FileSystem: file
        Path: my_ts_opt_dataset.zarr
        Hidden: True

      - ...

.. _config:
Server Demo Configuration File
==============================
The server configuration file consists of various parts, some of them are necessary others are optional.

> Here the `demo configuration file`_ used in the `example`_ is explained in detail.

Reviewer: The example link does not work as you have two examples in this file. Can you rename them appropriately?

> The configuration file consists of five main parts `authentication`_, `dataset attribution`_, `datasets`_,
> `place groups`_ and `styles`_.

Reviewer: ```suggestion
The configuration file consists of five main parts `authentication-optional`_, `dataset attribution-optional`_, `datasets-mandatory`_,
`place groups-optional`_ and `styles-optional`_.
```

The above example of how to specify a xcube dataset to be served above is using a datacube stored in
an S3 bucket within the Amazon Cloud. Please have a closer look to the parameter *Anonymous: true*.
This means, the datasets permissions are set to public read. If you have a dataset that is not public-public read, set
*Anonymous: false*. Furthermore, you need to have valid credentials on the machine where the server runs.
Credentials may be saved either in a file called .aws/credentials with content like below:

    [default]
    aws_access_key_id=AKIAIOSFODNN7EXAMPLE

>     aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY

Author: They are :)

The above example of how to specify a xcube dataset to be served above is using a datacube stored in
an S3 bucket within the Amazon Cloud. Please have a closer look at the parameter *Anonymous: true*.
This means, the datasets permissions are set to public read in your source s3 bucket. If you have a dataset that is not public-read, set
*Anonymous: false*. Furthermore, you need to have valid credentials on the machine where the server runs.
Credentials may be saved either in a file called .aws/credentials with content like below:

    | [default]
    | aws_access_key_id=AKIAIOSFODNN7EXAMPLE
    | aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY

Or they may be exported as environment variables AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID.

Further down an example for a `locally-stored-xcube-datasets`_ will be given,

> as well as an example of a `on-the-fly-generation-of-xcube-datasets`_.

Reviewer: Links are not working.

*BoundingBox* [optional]
may be set in order to restrict the region which is served from a certain datacube. The
notation of the *BoundingBox* is [lon_min,lat_min,lon_max,lat_max].

*FileSystem* [mandatory]
is set to "s3" which lets xcube serve know, that the datacube is located in the cloud.

*Endpoint* [mandatory]
contains information about the cloud provider endpoint, this will differ if you use a different
cloud provider.

*Path* [mandatory]
leads to the specific location of the datacube. The particular datacube is stored in an
OpenTelecomCloud S3 bucket called "xcube-examples" and the datacube is called "OLCI-SNS-RAW-CUBE-2.zarr".

*Region* [optional]
is the region where the specified cloud provider is operating.

*Styles* [optional]
influence the visualization of the xucbe dataset in the xcube viewer if specified in the server configuration file. The usage of *Styles* is described in section `styles`_.

*PlaceGroups* [optional]
allow to associate places (e.g. polygons or point-location) with a particular xcube dataset.
Several different place groups may be connected to a xcube dataset, these different place groups are distinguished by

> the *PlaceGroupRef*. The configuration of *PlaceGroups* is described in section `place groups`_.

Reviewer: ```suggestion
the *PlaceGroupRef*. The configuration of *PlaceGroups* is described in section `place-groups-optional`_.
```

*AccessControl* [optional]

> can only be used when providing `authentication`_. Datasets may be protected by

Reviewer: ```suggestion
can only be used when providing `authentication-optional`_. Datasets may be protected by
```

    [I 190924 17:08:54 service:159] press CTRL+C to stop service

Server Demo Configuration File for DataStores
=============================================

> The server configuration file consists of various parts, some of them are necessary others are optional.

Reviewer: ```suggestion
The server configuration file consists of various parts, some of them are necessary, others are optional.
```

Here the `demo_stores configuration file`_ used in the :ref:`example-stores` is explained in detail.

> This configuration file differs only in one part compared to :ref:`..config`. This part is :ref:`data-stores`,
> the other main parts :ref:`authentication`, :ref:`dataset-attribution`,
> :ref:`place-groups` and :ref:`styles`. Can be used in combination with :ref:`data-stores`.

Reviewer: ```suggestion
This configuration file differs only in one part compared from :ref:`..config`. This part is :ref:`data-stores`.
The other main parts (:ref:`authentication`, :ref:`dataset-attribution`,
:ref:`place-groups`, and :ref:`styles`) can be used in combination with :ref:`data-stores`.
```

.. _data-stores:
DataStores [mandatory]
--------------------

Datasets, which are stored in the same location, may be configured in the configuration file using *DataStores*.

.. code:: yaml

    DataStores:
      - Identifier: edc
        StoreId: s3
        StoreParams:
          root: xcube-dcfs/edc-xc-viewer-data
          max_depth: 1
          storage_options:
            anon: true
            # client_kwargs:
            #  endpoint_url: https://s3.eu-central-1.amazonaws.com
        Datasets:
          - Path: "*2.zarr"
            Style: "default"
            # ChunkCacheSize: 1G

*Identifier* [mandatory] is a unique ID for each DataStore.

*StoreID* [mandatory] can be *file* for locally stored datasets and *s3* for datasets located in the cloud.

*StoreParams* [mandatory]

> *root* [mandatory] gives the common part of the path of the served datasets.

Reviewer: ```suggestion
*root* [mandatory] states a common beginning part of the paths of the served datasets.
```

> *max_depth* [optional] if wildcard is used in *Dataset Path* this indicated, how far the server should step down and
> serve the discovered datasets.

Reviewer: ```suggestion
*max_depth* [optional] if wildcard is used in *Dataset Path* this indicates how far the server should step down in the directory structure to detect datasets that shall be served.
```

*storage_options* [optional] is necessary when serving datasets from the cloud. With *anon* the accessibility is

> configured, if the datasets are public-read, *anon* is set to "true", "false" indicates they are protected.

Reviewer: Add "Credentials" my be set with keywords 'key' and 'secret'."

*Datasets* [optional] if not specified, every dataset in the indicated location supported by xcube will be read and
served by xcube serve. In order to filter certain datasets you can list Paths that shall be served by xcube serve.
*Path* may contain wildcards. Each Dataset entry may have *Styles* and *PlaceGroups* associated with them, the same way

> as in :ref:`..config`.

Reviewer: Have a look at the layout on rtd

    [I 190924 17:08:54 service:159] press CTRL+C to stop service

Server Demo Configuration File for DataStores
=============================================
The server configuration file consists of various parts, some of them are necessary, others are optional.
Here the `demo stores configuration file`_ used in the `example stores`_ is explained in detail.

> This configuration file differs only in one part compared from `config`_. This part is `data stores`_.

Reviewer: ```suggestion
This configuration file differs only in one part from :ref:`config`.
`the configuration above`_: `data stores`_.
```

Most of the configuration of locally stored datasets is equal to the configuration of
:ref:`remotely-stored-xcube-datasets`.
`remotely stored xcube datasets`_.
